ctlcmd/cmdbase.py

```diff
@@ -771,7 +773,189 @@ def globcomp(text):
+                       default=self.DEFAULT_ROOTFILE,
+                       help="""File path to save (placeholders in angle braces).
+                       default=%(default)s""")
+ 
+  def parse(self,args):
+    if not args.saveroot:  # Early exit if savefile is not set
+      self.saveroot = None
+      return args
+    filename = args.saveroot
+
+    # Adding time stamp filenames
+    timestring = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')
+    filename = re.sub('<TIMESTAMP>', timestring, filename, flags=re.IGNORECASE)
+
+    # Adding boardid to the settings
+    filename = re.sub('<BOARDID>',
+                      str(self.board.boardid),
+                      filename,
+                      flags=re.IGNORECASE)
+    # Adding boardtype to the settings
+    filename = re.sub('<BOARDTYPE>',
+                      str(self.board.boardtype),
+                      filename,
+                      flags=re.IGNORECASE)
 
+    # Substituting tokens for argument values
+    for action in self.parser._actions:
+      for string in action.option_strings:
+        string = string.strip('-')
+        if not hasattr(args, string): continue
+        val = ''
+        try:
+          val = action.type(getattr(args, string))
+        except:
+          pass
+        substring = '{0}{1}'.format(string, val)
+        substring = re.sub(r'\.', 'p', substring)
+        filename = re.sub('\\<{0}\\>'.format(string),
+                          substring,
+                          filename,
+                          flags=re.IGNORECASE)
+    self.saveroot = filename
+    return args
+      
+  def maketree(self,data,datatypes={}):
+    """
+    @brief Create the root file and the TTree in the root file
+
+    @details This command creates the root file and creates the TTree with the correct
+    data types and names for each branch of the tree. This also saves some extra basic data 
+    like the time the file was created, the Board_ID and Board_Type. This function is only called once.
+
+    """
+     
+    self.rootfile = uproot.recreate(self.saveroot)
+    standard_dict={"time":np.float32,"det_id":int,"gantry x":np.float32,"gantry y":np.float32,
+                   "gantry z":np.float32,"LED bias voltage":np.float32,"LED temp":np.float32,"SiPM temp":np.float32} 
+    self.specific_dict={**{title:type(data[title]) for title in data}}
+
+    ##manual override if type is not returning a value that root will accept
+    if datatypes:
+      for title in datatypes:
+        self.specific_dict[title]=datatypes[title]
+    
+    self.rootfile.mktree("DataTree",{**{title:standard_dict[title] for title in standard_dict},**{title:self.specific_dict[title] for title in self.specific_dict}})
+
+    ##add extra data which must only be saved once
+    timestring = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')
+    self.rootfile["Board_ID"]=self.board.boardid
+    self.rootfile["Board_Type"]=self.board.boardtype
+    self.rootfile["Time"]=timestring
```

Comment:
We might want to group this into a separate tree: 

So 
```python
self.rootfile["cmd_stamp"]["board_id"] = self.board.boardid
self.rootfile["cmd_stamp"]["timestamp"] = timestamp
```
and such

(Thinking about what might be the best name for this rather than `cmd_stamp`. Since this should be the unique identifier for the command.)

Reply:
It appears that uproot complains when you upload strings into a ttree. I made a folder instead so that it will complain less but it is still sorted.